File: src/behaviorsense/agents/reasoning/verifier.py
"""Agent 4 - claim-level faithfulness verification.

This module is the project's primary research contribution.

The problem: "an LLM explains the alerts" is a demo, not research, because the obvious
question - *how do you know the explanation is true?* - has no answer. LLMs asked to
narrate numeric state routinely invent figures, invert directions ("increased" for a
decrease), and cite evidence that does not exist. In a caregiver-facing health context
that is not a cosmetic flaw.

The approach: constrain generation to a schema in which every claim must carry a
machine-resolvable `evidence_ref`, then verify each claim ARITHMETICALLY against the
behavioural state store. Four independent checks per claim:

    C1  ref_exists          does the cited evidence row exist at all?
    C2  value_matches       does the quoted number match the stored value?
    C3  pct_matches         does the quoted percentage match a recomputation?
    C4  direction_consistent  does the prose direction match the sign of the delta?

C4 is the subtle one and catches the most dangerous failure: a report that quotes the
right number while describing it backwards ("mobility improved by 42%" when it fell)
is worse than no report, and C1-C3 alone would pass it.

Design constraint - THE VERIFIER USES NO LLM. Using a model to check a model shares
failure modes and inherits the same blind spots; arithmetic does not. Every check here
is deterministic, cheap, and independently auditable.

Output metric: `hallucination_rate` = fraction of claims failing any check. Reported
constrained vs unconstrained, this is the headline experimental result.
"""

from __future__ import annotations

import re
from collections.abc import Iterable, Sequence
from dataclasses import dataclass, field

from behaviorsense.schemas import (
    BehaviourState,
    Claim,
    Evidence,
    VerificationResult,
)

REF_PATTERN = re.compile(r"^feat:([a-z0-9_]+):(\d{4}-\d{2}-\d{2})$")

# Direction words that may appear in claim prose. Used by the lexical cross-check,
# which catches text/field disagreement inside a single claim.
INCREASE_WORDS = frozenset({
    "increase", "increased", "increases", "rose", "rise", "risen", "higher",
    "up", "improved", "improvement", "gained", "more", "longer", "greater",
    "above", "exceeded", "grew",
})
DECREASE_WORDS = frozenset({
    "decrease", "decreased", "decreases", "fell", "fall", "fallen", "lower",
    "down", "declined", "decline", "reduced", "reduction", "less", "fewer",
    "shorter", "below", "dropped", "drop", "worsened", "deteriorated",
})

_WORD_RE = re.compile(r"[a-z]+")


@dataclass(frozen=True)
class VerifierConfig:
    """Tolerances for numeric agreement.

    Tolerances are deliberately tight. A generous tolerance would let a model paraphrase
    numbers ("about 40%") into a pass and inflate the faithfulness score, which would
    quietly destroy the metric's meaning.
    """

    value_rel_tol: float = 0.02
    """2% relative tolerance on quoted absolute values. Accommodates sensible rounding
    (1800.0 -> "1800" or "30 minutes") without admitting invention."""

    value_abs_tol: float = 1e-6
    """Absolute floor for near-zero comparisons where relative tolerance is meaningless."""

    pct_abs_tol: float = 2.0
    """Percentage points. A claim of "-42%" against a true -40.5% passes; "-25%" fails."""

    require_evidence_ref: bool = True
    """If False, claims without refs are counted unverifiable rather than unfaithful.
    Used to score unconstrained-baseline output fairly in the ablation."""

    check_lexical_direction: bool = True
    """Cross-check direction words in `text` against the structured `direction` field."""


@dataclass
class VerificationReport:
    """Aggregate verification outcome over a set of claims."""

    results: list[VerificationResult] = field(default_factory=list)
    n_claims: int = 0

    @property
    def n_faithful(self) -> int:
        return sum(1 for r in self.results if r.is_faithful)

    @property
    def hallucination_rate(self) -> float:
        """Fraction of claims failing at least one check. The headline metric."""
        if self.n_claims == 0:
            return 0.0
        return 1.0 - (self.n_faithful / self.n_claims)

    @property
    def faithfulness_rate(self) -> float:
        return 1.0 - self.hallucination_rate

    def failure_breakdown(self) -> dict[str, int]:
        """Per-check failure counts, for the error analysis table."""
        return {
            "missing_or_bad_ref": sum(1 for r in self.results if not r.ref_exists),
            "value_mismatch": sum(
                1 for r in self.results if r.ref_exists and not r.value_matches
            ),
            "pct_mismatch": sum(
                1 for r in self.results if r.ref_exists and not r.pct_matches
            ),
            "direction_error": sum(
                1 for r in self.results if r.ref_exists and not r.direction_consistent
            ),
        }

    def summary(self) -> str:
        b = self.failure_breakdown()
        return (
            f"claims={self.n_claims} faithful={self.n_faithful} "
            f"hallucination_rate={self.hallucination_rate:.3f} | "
            f"bad_ref={b['missing_or_bad_ref']} value={b['value_mismatch']} "
            f"pct={b['pct_mismatch']} direction={b['direction_error']}"
        )


class FaithfulnessVerifier:
    """Deterministic claim verifier.

    Usage::

        verifier = FaithfulnessVerifier()
        report = verifier.verify_all(claims, state)
        print(report.hallucination_rate)
    """

    def __init__(self, config: VerifierConfig | None = None) -> None:
        self.config = config or VerifierConfig()

    # -- evidence resolution ----------------------------------------------------

    def build_index(self, state: BehaviourState) -> dict[str, Evidence]:
        """Index every resolvable evidence ref for this state.

        Alert evidence is the primary source, but a claim may legitimately cite any
        tracked feature for the report day - the LLM is shown the full daily feature
        vector, not only alerting features. Synthesising evidence for non-alerting
        features keeps honest contextual claims ("meals were normal at 3") verifiable
        instead of being wrongly scored as hallucinations.
        """
        index: dict[str, Evidence] = dict(state.evidence_index())

        day = state.report_day
        items = state.today.numeric_items()
        for name, value in items.items():
            ref = f"feat:{name}:{day.isoformat()}"
            if ref in index:
                continue
            baseline = state.baselines.get(name)
            if baseline is None:
                continue
            index[ref] = Evidence.build(name, value, baseline, day)
        return index

    # -- individual checks ------------------------------------------------------

    def _values_agree(self, claimed: float, actual: float) -> bool:
        cfg = self.config
        if abs(actual) < cfg.value_abs_tol:
            return abs(claimed) < max(cfg.value_abs_tol, 1e-3)
        return abs(claimed - actual) <= cfg.value_rel_tol * abs(actual)

    def _lexical_direction(self, text: str) -> str | None:
        """Infer direction from prose. None if absent or contradictory.

        Contradictory text (both increase and decrease words) yields None rather than a
        guess - an ambiguous sentence should not be scored as a confident pass.
        """
        words = set(_WORD_RE.findall(text.lower()))
        up = bool(words & INCREASE_WORDS)
        down = bool(words & DECREASE_WORDS)
        if up == down:  # neither, or both
            return None
        return "increase" if up else "decrease"

    def verify_claim(
        self, claim: Claim, index: dict[str, Evidence]
    ) -> VerificationResult:
        cfg = self.config
        notes: list[str] = []

        # C1 - reference resolution.
        if not claim.evidence_ref:
            return VerificationResult(
                claim_id=claim.claim_id,
                ref_exists=not cfg.require_evidence_ref,
                value_matches=False,
                pct_matches=False,
                direction_consistent=False,
                notes=["no evidence_ref supplied"],
            )

        if not REF_PATTERN.match(claim.evidence_ref):
            return VerificationResult(
                claim_id=claim.claim_id,
                ref_exists=False,
                value_matches=False,
                pct_matches=False,
                direction_consistent=False,
                notes=[f"malformed evidence_ref: {claim.evidence_ref!r}"],
            )

        evidence = index.get(claim.evidence_ref)
        if evidence is None:
            return VerificationResult(
                claim_id=claim.claim_id,
                ref_exists=False,
                value_matches=False,
                pct_matches=False,
                direction_consistent=False,
                notes=[
                    f"evidence_ref {claim.evidence_ref!r} does not resolve "
                    "(fabricated citation)"
                ],
            )

        # C2 - quoted absolute value.
        if claim.claimed_value is None:
            value_matches = True
            notes.append("no absolute value quoted; C2 vacuous")
        else:
            value_matches = self._values_agree(
                claim.claimed_value, evidence.observed_value
            )
            if not value_matches:
                notes.append(
                    f"value mismatch: claimed {claim.claimed_value:g}, "
                    f"actual {evidence.observed_value:g}"
                )

        # C3 - quoted percentage change.
        if claim.claimed_pct_change is None:
            pct_matches = True
            notes.append("no percentage quoted; C3 vacuous")
        elif abs(evidence.baseline_median) < 1e-9:
            # Percentage change is undefined against a zero baseline; quoting one is
            # itself a defect, since no correct value exists.
            pct_matches = False
            notes.append(
                "percentage quoted against a zero baseline median (undefined)"
            )
        else:
            pct_matches = (
                abs(claim.claimed_pct_change - evidence.pct_change)
                <= cfg.pct_abs_tol
            )
            if not pct_matches:
                notes.append(
                    f"pct mismatch: claimed {claim.claimed_pct_change:+.1f}%, "
                    f"actual {evidence.pct_change:+.1f}%"
                )

        # C4 - direction. The check that catches inverted narration.
        actual_direction = evidence.direction
        direction_consistent = True

        if claim.direction is not None and claim.direction != actual_direction:
            direction_consistent = False
            notes.append(
                f"direction field wrong: claimed {claim.direction}, "
                f"actual {actual_direction} (delta {evidence.delta:+g})"
            )

        if cfg.check_lexical_direction:
            lexical = self._lexical_direction(claim.text)
            if lexical is not None and actual_direction != "unchanged":
                if lexical != actual_direction:
                    direction_consistent = False
                    notes.append(
                        f"prose says {lexical!r} but data shows {actual_direction!r} "
                        f"(delta {evidence.delta:+g})"
                    )
            # Internal contradiction between the structured field and the prose.
            if (
                lexical is not None
                and claim.direction is not None
                and lexical != claim.direction
            ):
                direction_consistent = False
                notes.append(
                    f"claim is self-inconsistent: text implies {lexical!r}, "
                    f"direction field says {claim.direction!r}"
                )

        # A quoted percentage whose sign contradicts the data is a direction error too,
        # even when the model omitted the direction field and used neutral prose.
        if (
            claim.claimed_pct_change is not None
            and abs(claim.claimed_pct_change) > 1e-9
            and abs(evidence.delta) > 1e-9
        ):
            if (claim.claimed_pct_change > 0) != (evidence.delta > 0):
                direction_consistent = False
                notes.append(
                    f"quoted pct sign ({claim.claimed_pct_change:+.1f}%) contradicts "
                    f"actual delta ({evidence.delta:+g})"
                )

        return VerificationResult(
            claim_id=claim.claim_id,
            ref_exists=True,
            value_matches=value_matches,
            pct_matches=pct_matches,
            direction_consistent=direction_consistent,
            notes=notes,
        )

    # -- batch ------------------------------------------------------------------

    def verify_all(
        self, claims: Sequence[Claim], state: BehaviourState
    ) -> VerificationReport:
        index = self.build_index(state)
        results = [self.verify_claim(c, index) for c in claims]
        return VerificationReport(results=results, n_claims=len(claims))

    def filter_faithful(
        self, claims: Sequence[Claim], state: BehaviourState
    ) -> tuple[list[Claim], VerificationReport]:
        """Return only claims that pass every check, plus the full report.

        This is the deployment path: unfaithful claims are withheld from the caregiver
        rather than shown with a warning. A flagged-but-visible false clinical claim is
        still a false clinical claim.
        """
        report = self.verify_all(claims, state)
        ok = {r.claim_id for r in report.results if r.is_faithful}
        return [c for c in claims if c.claim_id in ok], report


def aggregate_reports(reports: Iterable[VerificationReport]) -> VerificationReport:
    """Pool many per-report verifications into one corpus-level result."""
    pooled = VerificationReport()
    for r in reports:
        pooled.results.extend(r.results)
        pooled.n_claims += r.n_claims
    return pooled


__all__ = [
    "REF_PATTERN",
    "VerifierConfig",
    "VerificationReport",
    "FaithfulnessVerifier",
    "aggregate_reports",
]

File: web/scripts/verifier.js
/* The verifier: C1-C5, deterministic, no language model.
 *
 * A faithful port of `src/behaviorsense/agents/reasoning/verifier.py` — same tolerances,
 * same word lists, same order of checks, same treatment of a percentage against a zero
 * baseline. It runs here so the report at the top of the page is really being verified
 * rather than showing a pre-baked answer.
 *
 * C5 is the newest check and the one the measured run made the case for. C2 compares the
 * structured `claimed_value` against the evidence row, so it is blind to a claim whose row
 * is right and whose SENTENCE is not — and the sentence is what a caregiver reads. In the
 * latest run C2 was zero and C5 was the largest failure category (9 of 16 constrained, 13
 * of 20 free): the model copies the figure into the field correctly, then paraphrases it
 * in the prose. Without this port the page's agreement banner would report drift against
 * the server on exactly those claims, so C5 ships here the moment it shipped in Python.
 *
 * One deliberate difference, and it is presentational only. When a citation does not
 * resolve, Python returns C2-C5 as `false` because there is nothing to compare against.
 * Reporting five failures for one fault would overstate what went wrong, so unreachable
 * checks are marked `skip` here and named as such in the margin. The verdict — the claim
 * is unfaithful and is withheld — is identical.
 */

export const REF_PATTERN = /^feat:[a-z0-9_]+:\d{4}-\d{2}-\d{2}$/;

/* Notes are HTML by design — `<b>` marks the numbers being compared — so anything
 * interpolated into them has to be escaped, and the fields that need it are exactly the
 * ones the model chose: `evidence_ref` and `direction`. This mattered the moment the page
 * started rendering LIVE claims: they arrive from an unauthenticated tunnel, having been
 * written by a hosted language model, and a claim body reading `<img onerror=…>` is not a
 * hypothetical for text nobody wrote by hand. Exported so the ledger uses the same one.
 */
export const escapeHtml = (s) => String(s).replace(/[&<>"']/g, (c) =>
  ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[c]));

/* displayHalfStep mirrors verifier.py: the reporter publishes values rounded to 2
 * decimals, and a digit-for-digit copy of one must verify - 2% of a small value
 * (mobility_index 0.074) is tighter than the rounding itself. */
const TOL = { valueRel: 0.02, valueAbs: 1e-6, pctAbs: 2.0, displayHalfStep: 0.005 };

const INCREASE = new Set(["increase", "increased", "increases", "rose", "rise", "risen",
  "higher", "up", "improved", "improvement", "gained", "more", "longer", "greater",
  "above", "exceeded", "grew"]);
const DECREASE = new Set(["decrease", "decreased", "decreases", "fell", "fall", "fallen",
  "lower", "down", "declined", "decline", "reduced", "reduction", "less", "fewer",
  "shorter", "below", "dropped", "drop", "worsened", "deteriorated"]);

const directionOf = (delta) =>
  Math.abs(delta) < 1e-9 ? "unchanged" : delta > 0 ? "increase" : "decrease";

/* Every numeric literal in prose, for C5. The comma form is matched FIRST in the
 * alternation so "1,800" reads as one number — a scan that matched "1" then "800" was the
 * bug the Python port hit first, and the fix is the same shape here: match the grouped
 * form, then strip the commas before parsing. The k/M scale suffix needs a negative
 * lookahead so "5 minutes" does not read as 5,000,000. */
const NUM_RE = /-?\d{1,3}(?:,\d{3})+(?:\.\d+)?(?:[kKmM](?![a-zA-Z0-9]))?|-?\d+(?:\.\d+)?(?:[kKmM](?![a-zA-Z0-9]))?/g;

function numbersInText(text) {
  return [...String(text).matchAll(NUM_RE)].map((m) => {
    let s = m[0].replace(/,/g, "");
    let scale = 1;
    const suffix = s.slice(-1);
    if (suffix === "k" || suffix === "K") { scale = 1e3; s = s.slice(0, -1); }
    else if (suffix === "m" || suffix === "M") { scale = 1e6; s = s.slice(0, -1); }
    const v = parseFloat(s);
    return Number.isFinite(v) ? v * scale : null;
  }).filter((v) => v !== null);
}

/* C5 — does the prose echo the quoted value? Same 2% relative tolerance as C2, applied
 * symmetrically: a caregiver says "1,800 s" or "1.8k", and neither is a misquote. A claim
 * with no `claimed_value` has nothing to echo, so C5 is vacuous rather than failing. */
function proseQuotesValue(text, value) {
  for (const n of numbersInText(text)) {
    if (Math.abs(value) < TOL.valueAbs) {
      if (Math.abs(n) < Math.max(TOL.valueAbs, 1e-3)) return true;
      continue;
    }
    if (Math.abs(n - value) <= Math.max(TOL.valueRel * Math.abs(value),
                                        TOL.displayHalfStep)) return true;
  }
  return false;
}

function valuesAgree(claimed, actual) {
  if (Math.abs(actual) < TOL.valueAbs) return Math.abs(claimed) < Math.max(TOL.valueAbs, 1e-3);
  return Math.abs(claimed - actual) <= Math.max(TOL.valueRel * Math.abs(actual),
                                                TOL.displayHalfStep);
}

/** Direction implied by the prose. Null when absent or self-contradictory: an ambiguous
 * sentence must not be scored as a confident pass. */
function lexicalDirection(text) {
  const words = new Set((text.toLowerCase().match(/[a-z]+/g) ?? []));
  let up = false, down = false;
  for (const w of words) {
    if (INCREASE.has(w)) up = true;
    if (DECREASE.has(w)) down = true;
  }
  if (up === down) return null;
  return up ? "increase" : "decrease";
}

const fmt = (n) => {
  const abs = Math.abs(n);
  const digits = abs >= 100 || Number.isInteger(n) ? 0 : 1;
  return n.toLocaleString("en-GB", { minimumFractionDigits: digits, maximumFractionDigits: digits });
};
const pct = (n) => `${n >= 0 ? "+" : "−"}${Math.abs(n).toFixed(1)}%`;

/**
 * Verify one claim against the evidence index.
 * @returns {{checks: Array<{id: string, state: "pass"|"fail"|"skip"}>, faithful: boolean,
 *            notes: string[]}}
 */
export function verifyClaim(claim, index) {
  const notes = [];
  const mk = (c1, c2, c3, c4, c5) => [
    { id: "C1", state: c1 }, { id: "C2", state: c2 },
    { id: "C3", state: c3 }, { id: "C4", state: c4 }, { id: "C5", state: c5 },
  ];

  // C1 — reference resolution.
  if (!claim.evidence_ref) {
    return { checks: mk("fail", "skip", "skip", "skip", "skip"), faithful: false,
      notes: ["No evidence reference supplied, so nothing can be checked."] };
  }
  if (!REF_PATTERN.test(claim.evidence_ref)) {
    return { checks: mk("fail", "skip", "skip", "skip", "skip"), faithful: false,
      notes: [`Malformed reference <b>${escapeHtml(claim.evidence_ref)}</b>.`] };
  }
  const e = index[claim.evidence_ref];
  if (!e) {
    return { checks: mk("fail", "skip", "skip", "skip", "skip"), faithful: false,
      notes: [`Cites <b>${escapeHtml(claim.evidence_ref.split(":")[1])}</b>, which was not `
        + "measured today — a fabricated citation. C2 to C5 cannot be evaluated."] };
  }

  // C2 — quoted absolute value.
  let c2 = "pass";
  if (claim.claimed_value == null) {
    notes.push("No value quoted, so C2 is vacuous.");
  } else if (!valuesAgree(claim.claimed_value, e.observed_value)) {
    c2 = "fail";
    notes.push(`Quotes <b>${fmt(claim.claimed_value)}</b>, recorded value is `
      + `<b>${fmt(e.observed_value)}</b>.`);
  }

  // C3 — quoted percentage change.
  let c3 = "pass";
  if (claim.claimed_pct_change == null) {
    notes.push("No percentage quoted, so C3 is vacuous.");
  } else if (Math.abs(e.baseline_median) < 1e-9) {
    c3 = "fail";
    notes.push("Quotes a percentage against a zero baseline, where no correct value exists.");
  } else if (Math.abs(claim.claimed_pct_change - e.pct_change) > TOL.pctAbs) {
    c3 = "fail";
    notes.push(`Quotes <b>${pct(claim.claimed_pct_change)}</b>, recomputes to `
      + `<b>${pct(e.pct_change)}</b>.`);
  }
  return finishDirection(claim, e, mk, c2, c3, notes);
}

/* C4 — the check that catches inverted narration, and the reason C1-C3 are not enough.
 * Four independent ways a claim can tell the wrong story with the right number: the
 * structured field disagrees with the data, the prose disagrees with the data, the prose
 * disagrees with the field, or a quoted percentage carries the wrong sign. C5 then asks
 * the last question: does the sentence the caregiver actually reads contain the figure
 * the field records? */
function finishDirection(claim, e, mk, c2, c3, notes) {
  const actual = directionOf(e.delta);
  const lexical = lexicalDirection(claim.text);
  let c4 = "pass";

  if (claim.direction != null && claim.direction !== actual) {
    c4 = "fail";
    notes.push(`States <b>${escapeHtml(claim.direction)}</b>; the change is `
      + `<b>${actual}</b> (Δ ${e.delta > 0 ? "+" : "−"}${fmt(Math.abs(e.delta))}).`);
  }
  if (lexical != null && actual !== "unchanged" && lexical !== actual) {
    c4 = "fail";
    notes.push(`The sentence reads as <b>${lexical}</b>; the data show <b>${actual}</b>.`);
  }
  if (lexical != null && claim.direction != null && lexical !== claim.direction) {
    c4 = "fail";
    notes.push(`Self-inconsistent: prose implies <b>${lexical}</b>, field says `
      + `<b>${escapeHtml(claim.direction)}</b>.`);
  }
  if (claim.claimed_pct_change != null && Math.abs(claim.claimed_pct_change) > 1e-9
      && Math.abs(e.delta) > 1e-9
      && (claim.claimed_pct_change > 0) !== (e.delta > 0)) {
    c4 = "fail";
    notes.push(`Quoted percentage sign contradicts the measured change.`);
  }

  let c5 = "pass";
  if (claim.claimed_value == null) {
    notes.push("No value quoted, so C5 is vacuous.");
  } else if (!proseQuotesValue(claim.text, claim.claimed_value)) {
    c5 = "fail";
    notes.push(`The sentence never quotes the <b>${fmt(claim.claimed_value)}</b> the field `
      + "records — a caregiver reading the prose would take away a different figure.");
  }

  const checks = mk("pass", c2, c3, c4, c5);
  const faithful = checks.every((c) => c.state !== "fail");
  return { checks, faithful, notes: faithful ? [] : notes.filter((n) => !n.includes("vacuous")) };
}

/** Verify a whole report. */
export function verifyAll(claims, index) {
  return claims.map((claim) => ({ claim, ...verifyClaim(claim, index) }));
}


